roles/jd-server/src/lib/job_declarator/mod.rs

```diff
@@ -100,6 +109,68 @@ impl JobDeclaratorDownstream {
+                                        (last_declare, tx_list, _x)
+                                    }
+                                    None => {
+                                        warn!("Received solution but no job available");
+                                        todo!()
+                                    }
+                                };
+                                let coinbase_pre = last_declare.coinbase_prefix.to_vec();
+                                let extranonce = message.extranonce.to_vec();
+                                let coinbase_suf = last_declare.coinbase_suffix.to_vec();
+                                let mut path: Vec<Vec<u8>> = vec![];
+                                for tx in &tx_list {
+                                    let id = tx.txid();
+                                    let id = id.as_ref().to_vec();
+                                    path.push(id);
+                                }
+                                let merkle_root = merkle_root_from_path(
+                                    &coinbase_pre[..],
+                                    &coinbase_suf[..],
+                                    &extranonce[..],
+                                    &path,
+                                )
+                                .expect("Invalid coinbase");
+                                let merkle_root = Hash::from_inner(merkle_root.try_into().unwrap());
+
+                                let prev_blockhash =
+                                    u256_to_block_hash(message.prev_hash.into_static());
+                                let header =
+                                    stratum_common::bitcoin::blockdata::block::BlockHeader {
+                                        version: last_declare.version as i32,
+                                        prev_blockhash,
+                                        merkle_root,
+                                        time: message.ntime,
+                                        bits: message.nbits,
+                                        nonce: message.nonce,
+                                    };
+
+                                let coinbase = [coinbase_pre, extranonce, coinbase_suf].concat();
+                                let coinbase = Transaction::deserialize(&coinbase[..]).unwrap();
+                                tx_list.insert(0, coinbase);
+
+                                let mut block = Block {
+                                    header,
+                                    txdata: tx_list.clone(),
+                                };
+
+                                block.header.merkle_root = block.compute_merkle_root().unwrap();
+
+                                let serialized_block = serialize(&block);
+                                let hexdata = hex::encode(serialized_block);
+
+                                let _ = submit_solution_sender.send(hexdata).await;
+                            }
```

Comment:
Why did you move the SubmitSolution logic here, instead of managing it in the handlers as it is now?

Reply:
because, as I remember, this would require to make the handler trait async, which is not trivial

Reply:
I try to move where it was before and see if I can fix

Reply:
there are sveral possible solution is moving where was before put that somewhere else but not here

Reply:
added a method to JobDeclaratorDoenstream, that actually relies on a new function in `roles-logic-sv2/utils`
`
pub fn submit_solution_to_block( last_declare: DeclareMiningJob,  mut tx_list: Vec<bitcoin::Transaction>, message: SubmitSolutionJd) -> bitcoin::Block {...}
`

Reply:
this function is now turned into a From for Block that takes a wrapper struct